module/zfs/vdev_indirect.c

```diff
@@ -904,7 +908,11 @@ vdev_obsolete_counts_are_precise(vdev_t *vd)
 	err = zap_lookup(vd->vdev_spa->spa_meta_objset, vd->vdev_top_zap,
 	    VDEV_TOP_ZAP_OBSOLETE_COUNTS_ARE_PRECISE, sizeof (val), 1, &val);
 
-	ASSERT(err == 0 || err == ENOENT);
+	if (err != 0 && err != ENOENT) {
+		vdev_dbgmsg(vd, "vdev_load: %s failed to retrieve obsolete "
+		    "counts from vdev ZAP [error=%d]", FTAG, err);
+		ASSERT3S(err, ==, ECKSUM);
```

Comment:
I don't think that this semantic change is safe.  Have you examined all callers to ensure that we can treat the counts as imprecise?  It looks like spa_vdev_remove_cancel_sync() would not behave correctly - it would leak a ref on SPA_FEATURE_OBSOLETE_COUNTS (and leave the VDEV_TOP_ZAP_OBSOLETE_COUNTS_ARE_PRECISE lying around).

Reply:
Semantically, other than a log message, we haven't actually changed anything for a production build.  If the `zap_lookup()` were to fail, and it certainly can, the return value would be the same.  If the callers can't already handle that, and I agree it looks like they can't, then these code paths have not been entirely safe since there were added.  The same looks to be true for `vdev_obsolete_sm_object()` and its callers.

My feeling is the best way to handle this would be to rework the interface and callers to acknowledge that things aren't as simple as it exists or it  does exist.  There are other possible legitimate but annoying failure scenarios.

The only other strictly safe option I can see would be to convert the ASSERTs to VERIFYs so at least we crash the system before any damage can be done.  Which isn't great from a user perspective.

I should also mention that I only started looking at this because it is a case not infrequently hit by the CI, which runs at least one test that performance an extreme pool rewind.

Reply:
You're right, this should be a VERIFY, not an ASSERT.  Or as you said, we'd need to change this to be able to return an error and make the callers handle that.

Reply:
I think whatever is decided, we should also change https://github.com/zfsonlinux/zfs/commit/e927fc8a522e1c0db89955cc555841aa23bbd634, either on this review, or we can open a bug and I can make that change as a follow-on.

Reply:
Let me see about reworking this to return an error and updating all the callers.  I'd rather not be forced to disable the ZTS tests which can hit this.